examples/cfq/syntax-based/node.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2020 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""This module is for constructing the syntax tree from the action sequence."""
+
+import re
+from typing import List
+from grammar import Grammar, GRAMMAR_STR
+from asg import generate_action_sequence
+from collections import deque
+from asg import Action, APPLY_RULE, GENERATE_TOKEN
+
+class Node:
+
+  def __init__(self, parent: 'Node', value: str):
+    self.parent = parent
+    self.value = value
+    self.children = []
+
+  def add_child(self, child: 'Node'):
+    self.children.append(child)
+
+
+def apply_action(frontier_nodes_stack: deque, action: Action, grammar: Grammar):
+  """Applies an action (apply rule or generate token). The action extends a
+  AST that is under construction by extending the stack of frontier nodes.
+  The function returns the extended stack."""
+  current_node: Node = frontier_nodes_stack.pop()
+  action_type, action_value = action
+  if action_type == GENERATE_TOKEN:
+    # Generate leaf node with token stored in the action.
+    child = Node(current_node, action_value)
+    current_node.add_child(child)
+  else:
+    new_frontier_nodes = []
+    rule_name = action_value
+    head, body = grammar.sub_rules[rule_name]
+    if head != current_node.value:
+      raise Exception('Invalid action. Got {} and expected {}'.format(
+          head, current_node.value))
+    rule_tokens = body.split()
+    for rule_token in rule_tokens:
+      match = re.match(r'\"(.*)\"', rule_token)
```

Comment:
Like I said in the design doc, I think that if we are using ASTs we can get rid of all these strings inside the grammar, and add them in a postprocessing step. Let's chat about this during our 1:1 tomorrow.

Reply:
will be done in follow-up PR, as discussed